### PR TITLE
[SALTO-1371] Remove unnecessary references to createRefWithElement

### PR DESCRIPTION
**Pull request by yanaise**
Remove creation of references where it's not needed. Only in-code effects.
---
---
_Release Notes_: 
---
_User Notifications_:

### CODE REVIEW

**Review by adamro — COMMENTED** [review]
There's a big downside in memory with doing this change for `BuiltinTypes`.
When using `createRefWithElement`, there's a logic that re-uses one ReferenceExpression per BuiltinType.

If we don't use it and use the constructor, we lose this logic and create a new ReferenceExpression per usage.

In order to not have this downside, we need to implement a similar mechanism when creating the References in the constructor and not through the func, or to keep using the `createRefWithElement`

**Comment by ori-moisis** [comment]
@adamro - can the constructor just use that function internally?
I think the main source of confusion (and noise) was due to the fact that all the calls to the constructors changed, I don't think we had an issue with the logic

@yanaise - wasn't there a ticket number associated with this?

**Comment by yanaise** [comment]
> @adamro - can the constructor just use that function internally?
> I think the main source of confusion (and noise) was due to the fact that all the calls to the constructors changed, I don't think we had an issue with the logic
> 
> @yanaise - wasn't there a ticket number associated with this?

Yeah, I'll add the ticket number, sorry, and I spoke with Adam and indeed the constructor will use the function internally

File: packages/adapter-api/src/builtins.ts [hunk 1]
```diff
@@ -152,10 +152,15 @@ export const BuiltinTypesByFullName: Record<string, PrimitiveType> = (_.keyBy(
...
-  BuiltinTypesByFullName,
-  type => new ReferenceExpression(type.elemID, type)
-)
+// This is a pretty big hack: the map is created in elements, because it is used in
+// the Element constructor. But it can only be initialized here, after types have been defined.
+// Specifically, PrimitiveType needs to be defined before this is initialized, and PrimitiveType
+// inherits Element. To solve this, we would need to unite three modules: builtins, values,
+// and element. Even then, we would need to initialize an empty record, and only add values into it
+// after all necessary classes have been defined.
+Object.entries(BuiltinTypesByFullName).forEach(([name, type]) => {
```
Comment (adamro) [inline packages/adapter-api/src/builtins.ts:161]: Not sure what I think of this :)
I understand why it's needed, and I guess it's better than merging all the modules. So I guess it's a lesser evil.

Comment (yanaise) [reply packages/adapter-api/src/builtins.ts:161]: It sucks, but I spent a day trying to get around this, and there doesn't seem to be a way other than merging the modules

**Review by adamro — APPROVED** [review]
You need another rebase of course but approving to unblock you after you do that

**Pull request merged**